druid-shell/src/text_input.rs

```diff
@@ -0,0 +1,516 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::common_util::Counter;
+use crate::keyboard::{KbKey, KeyEvent};
+use crate::kurbo::{Point, Rect};
+use crate::piet::HitTestPoint;
+use crate::window::WinHandler;
+use std::borrow::Cow;
+use std::ops::Range;
+
+/// Uniquely identifies a text input field inside a window.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
+pub struct TextInputToken(u64);
+
+impl TextInputToken {
+    /// A token that does not correspond to any text input.
+    pub const INVALID: TextInputToken = TextInputToken(0);
+
+    /// Create a new token; this should for the most part be called only by platform code.
+    pub fn next() -> TextInputToken {
+        static TEXT_FIELD_COUNTER: Counter = Counter::new();
+        TextInputToken(TEXT_FIELD_COUNTER.next())
+    }
+
+    /// Create a new token from a raw value.
+    pub const fn from_raw(id: u64) -> TextInputToken {
+        TextInputToken(id)
+    }
+
+    /// Get the raw value for a token.
+    pub const fn into_raw(self) -> u64 {
+        self.0
+    }
+}
+
+/// An event that indicates to the platform that some aspect of a text field may have changed.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum TextInputUpdate {
+    /// Indicates the value returned by `TextInputHandler::selection` may have changed.
+    SelectionChanged,
+
+    /// Indicates the values returned by one or more of these methods may have changed:
+    /// - `TextInputHandler::hit_test_point`
+    /// - `TextInputHandler::line_range`
+    /// - `TextInputHandler::bounding_box`
+    /// - `TextInputHandler::slice_bounding_box`
+    LayoutChanged,
+
+    /// Indicates any value returned from any of the TextInputHandler methods may have changed.
+    Reset,
+}
+
+/// Corresponds to a selection of some text. This also can represent a text input caret — the blinking vertical
+/// bar where text is to be inserted. A caret is a selection with zero length, where `anchor == extent`.
+/// Indices are always expressed in UTF-8 bytes, and must be between 0 and the document length, inclusive.
+///
+/// As an example, if the input caret is at the start of the document `hello world`, we would expect both `anchor`
+/// and `extent` to be `0`. If the user holds shift and presses the right arrow key five times, we would expect the
+/// word `hello` to be selected, the `anchor` to still be `0`, and the `extent` to now be `5`.
+#[derive(Clone, Default, PartialEq, Eq, Hash)]
+pub struct TextSelection {
```

Comment:
That would definitely be ideal! I think it will require we modify this to include last-horizontal-move-pixel-x positions, but perhaps an IME will some day want that information anyway. It does mean we'll have to more explicitly specify how pixel-x positions (is there a better name for that which you've seen) change in response to edits?